File: packages/react-native-vision-camera-resizer/ios/HybridResizerFactory.swift
//
//  HybridResizerFactory.swift
//  VisionCamera
//
//  Created by Marc Rousavy on 29.10.25.
//

import AVFoundation
import Metal
import NitroModules

class HybridResizerFactory: HybridResizerFactorySpec {
  func isAvailable() -> Bool {
    return MTLCreateSystemDefaultDevice() != nil
  }

  func createResizer(options: ResizerOptions) throws -> Promise<any HybridResizerSpec> {
    return Promise.async {
      // Initialize Metal resizer pipeline asynchronously
      return try HybridResizer(options: options)
    }
  }
}

File: packages/react-native-vision-camera-resizer/android/src/main/cpp/HybridResizerFactory.cpp
///
/// HybridResizerFactory.cpp
/// VisionCamera
/// Copyright © 2026 Marc Rousavy @ Margelo
///

#include "HybridResizerFactory.hpp"

#include "vulkan/VulkanResizerPipeline.hpp"

namespace margelo::nitro::camera::resizer {

HybridResizerFactory::HybridResizerFactory() : HybridObject(TAG) {}

bool HybridResizerFactory::isAvailable() {
  return vulkan::VulkanResizerPipeline::isSupported();
}

std::shared_ptr<Promise<std::shared_ptr<HybridResizerSpec>>> HybridResizerFactory::createResizer(const ResizerOptions& options) {
  return Promise<std::shared_ptr<HybridResizerSpec>>::async([=]() {
    // Create `HybridResizer` on a separate C++ Thread (pooled),
    // so that the Vulkan setup doesn't block the main JS Thread.
    return std::make_shared<HybridResizer>(options);
  });
}

} // namespace margelo::nitro::camera::resizer
